### PR TITLE
cache: implement single-flight caching in the inmemory cache

### PR DESCRIPTION
**Pull request by GiedriusS**
Implement a single-flight mode for the inmemory cache. Since our code
goes and does something when Fetch() misses, let's implement the
single-flight mode based on this assumption - wait for the results of
the previous request and use it.

https://pkg.go.dev/golang.org/x/sync/singleflight is not used to avoid
complexity - there is no single function that we can call to get results
of some operation. Potentially it could be wrapped around the `Bucket`
interface but then the caching/remote object storage interfaces would be
muddled together and I really think we should avoid that.

Tested on a cluster and with unit tests.

Signed-off-by: Giedrius Statkevičius <giedrius.statkevicius@vinted.com>

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

File: pkg/cache/inmemory.go [hunk 1]
```diff
@@ -51,8 +65,10 @@ type InMemoryCache struct {
 	hitsExpired prometheus.Counter
 	// The input cache value would be copied to an inmemory array
 	// instead of simply using the one sent by the caller.
-	added            prometheus.Counter
-	current          prometheus.Gauge
+	added             prometheus.Counter
+	current           prometheus.Gauge
+	singleflightsaved prometheus.Counter
```
Comment (bwplotka) [inline pkg/cache/inmemory.go:70]:
```suggestion
	sfSavedCalls prometheus.Counter
```

File: pkg/cache/inmemory.go [hunk 2]
```diff
@@ -299,7 +342,32 @@ func (c *InMemoryCache) Fetch(ctx context.Context, keys []string) map[string][]b
 	for _, key := range keys {
 		if b, ok := c.get(key); ok {
 			results[key] = b
+		} else if c.singleFlight {
```
Comment (bwplotka) [inline pkg/cache/inmemory.go:345]: I would suggest avoiding else and return above so we can have more readable code. See https://github.com/thanos-io/thanos/blob/main/docs/contributing/coding-style-guide.md#control-structure-prefer-early-returns-and-avoid-else

**Review by bwplotka — DISMISSED** [review]
Interesting approach. I think I like it generally. Have you tried run any benchmarks with it? (:

**Comment by GiedriusS** [comment]
I don't write benchmarks every day but this small benchmark that I wrote only shows a difference with high concurrency for the same keys:

```
BenchmarkInmemorySingleflight/inmemory_get_set_keys10000_conc1-16                    	     139	   7217798 ns/op	 4316280 B/op	   59475 allocs/op
BenchmarkInmemorySingleflight/inmemory_singleflight_get_set_keys10000_conc1-16       	     136	   7452180 ns/op	 4315249 B/op	   59473 allocs/op
BenchmarkInmemorySingleflight/inmemory_get_set_keys10000_conc5-16                    	      32	  34888344 ns/op	21581605 B/op	  297421 allocs/op
BenchmarkInmemorySingleflight/inmemory_singleflight_get_set_keys10000_conc5-16       	      32	  35949705 ns/op	21582620 B/op	  297430 allocs/op
BenchmarkInmemorySingleflight/inmemory_get_set_keys10000_conc10-16                   	      15	  72321593 ns/op	43164038 B/op	  594854 allocs/op
BenchmarkInmemorySingleflight/inmemory_singleflight_get_set_keys10000_conc10-16      	      14	  76983315 ns/op	43164684 B/op	  594859 allocs/op
```

This is expected since we now do some locking :thinking:

File: pkg/cache/inmemory.go [hunk 3]
```diff
@@ -289,6 +315,24 @@ func (c *InMemoryCache) reset() {
 func (c *InMemoryCache) Store(ctx context.Context, data map[string][]byte, ttl time.Duration) {
 	for key, val := range data {
 		c.set(key, val, ttl)
+		if !c.singleFlight {
+			continue
+		}
+
+		c.mu.Lock()
```
Comment (ianbillett) [inline pkg/cache/inmemory.go:322]: Is it ok to be sharing this mutex across the read and write path? The naive argument would say - might this actually increase the time it takes to respond as we're waiting around a bunch.

**Review by ianbillett — COMMENTED** [review]
This feels like a good optimisation to make :+1: 

The main questions I have is:
* Do we **need** this? Are we getting a big enough increase in performance to justify the non-trivial code overhead? I would hope that the benchmarks will tell us this concretely.

Aside from that it took me quite a while to figure out what is `single-flighting` and what does it do for us. In fact - I'm still not 100% sure :sweat_smile: It would be good to explicitly spell this out both in the PR and the code.

File: pkg/cache/inmemory.go [hunk 4]
```diff
@@ -299,7 +343,36 @@ func (c *InMemoryCache) Fetch(ctx context.Context, keys []string) map[string][]b
...
+		if c.subs[key] == nil {
+			c.subs[key] = &pubsub{originalCtx: ctx}
+			c.mu.Unlock()
+		} else {
+			if c.subs[key].originalCtx.Err() != nil {
+				c.subs[key] = &pubsub{originalCtx: ctx}
+			}
```
[the remark this answers is not in the document]
Comment (GiedriusS) [reply pkg/cache/inmemory.go:359]: Yep :+1:

File: pkg/cache/inmemory.go (hunk 3 above)
Comment (GiedriusS) [reply pkg/cache/inmemory.go:322]: Reading and writing from a `map` from more than one goroutine `panic`s the whole program hence this lock is needed

File: pkg/cache/inmemory_test.go [hunk 5]
```diff
@@ -15,6 +19,123 @@ import (
...
+
+	g.Go(func() error {
+		hits := c.Fetch(context.Background(), testKeys)
+		if len(hits) != 1 {
+			return errors.New("expected to have 1 hit")
+		}
+		return nil
+	})
+
+	time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (GiedriusS) [reply pkg/cache/inmemory_test.go:91]: To wait until the previous goroutine spawned and blocked on the Fetch() call. In practice, that will take only a few nano/milliseconds. It would be hard to implement some kind of synchronization primitive here because we are actually calling the caching code here hence a simple `time.Sleep()` has been added.

File: pkg/cache/inmemory_test.go [hunk 6]
```diff
@@ -15,6 +19,123 @@ import (
...
+max_size: 1MB
+max_item_size: 2KB
+single_flight: true
+`)
+
+	c, _ := NewInMemoryCache("test", log.NewNopLogger(), nil, conf)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	const testKey = "test"
```
[the remark this answers is not in the document]
Comment (GiedriusS) [reply pkg/cache/inmemory_test.go:115]: Why not? It's constant and doesn't change. Ideally this would be a const slice but no such things exist in Go (:

File: pkg/cache/inmemory_test.go [hunk 7]
```diff
@@ -15,6 +19,123 @@ import (
...
+		hits := c.Fetch(ctx, []string{testKey})
+
+		if len(hits) != 0 {
+			return errors.New("got hits")
+		}
+		return nil
+	})
+	cancel()
+
+	time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (GiedriusS) [reply pkg/cache/inmemory_test.go:134]: Yep, because we want to wait until c.Fetch() has been called. In practice, we should only wait for a millisecond or so but a second is here to be on the safe side.

**Comment by GiedriusS** [comment]
> This feels like a good optimisation to make 
> 
> The main questions I have is:
> 
> * Do we **need** this? Are we getting a big enough increase in performance to justify the non-trivial code overhead? I would hope that the benchmarks will tell us this concretely.
> 
> Aside from that it took me quite a while to figure out what is `single-flighting` and what does it do for us. In fact - I'm still not 100% sure  It would be good to explicitly spell this out both in the PR and the code.

From my tests, the in-memory cache blows away all other options with this improvement because Thanos Store will try harder to avoid expensive read operations which is what's causing the sluggishness. Of course, it doesn't work well with multiple nodes. Ideally, we'd have https://github.com/thanos-io/thanos/issues/2962 but the library there needs extra improvements. Also, you can see this in action by opening some dashboards with this feature. The counter jumps to a hundred and thousands very quickly.

In fact, single-flight functionality for caches (not just the simple, in-memory one) is so good that I would argue that this needs to be the default. But this PR adds a toggle switch for it and users who do not want this will have no impact on the performance. So, I would really like to see this get merged.

**Comment by GiedriusS** [comment]
@thanos-io/thanos-maintainers could you PTAL? I'd love to get this into the next release :(

**Review by ianbillett — COMMENTED** [review]
From a functional point of view - I'm confident that this change does what it says it does. I do still think we are missing some benchmarks to make it obvious beyond any reasonable doubt that the increase in complexity is worth the performance boost.

From a maintainability perspective, I think there is still some changes to be made. 

This is a non-trivial bit of code that has taken me ~30 mins on both reviews to understand and conceptualise how it fits into the wider read & write paths. If the code is easy to understand, people will maintain it and vice versa - what [if Giedrus was hit by a bus](https://legacy.python.org/search/hypermail/python-1994q2/1040.html)?

Personally, I would like to see extra context in the comments and _ideally_ a sequence diagram that explains why we are making this change.

I'm not going to block this PR as I'm still pretty inexperienced in the code base, but that is my 2c :hugs:

File: pkg/cache/inmemory.go [hunk 8]
```diff
@@ -299,7 +342,36 @@ func (c *InMemoryCache) Fetch(ctx context.Context, keys []string) map[string][]b
...
+			c.mu.Unlock()
+
+			select {
+			case b := <-respReceiver:
+				results[key] = b
+				c.sfSaved.Inc()
+			case <-ctx.Done():
+				return results
+			case <-originalCtx.Done():
+				continue
```
[the remark this answers is not in the document]
Comment (ianbillett) [reply pkg/cache/inmemory.go:371]: If so - let's be explicit about that - took me a while to figure that out.

File: pkg/cache/inmemory.go [hunk 9]
```diff
@@ -289,6 +314,24 @@ func (c *InMemoryCache) reset() {
 func (c *InMemoryCache) Store(ctx context.Context, data map[string][]byte, ttl time.Duration) {
```
[the remark this answers is not in the document]
Comment (GiedriusS) [reply pkg/cache/inmemory.go:314]: The answer probably is that popular remote caches do not have such functionality - as far as I can tell from my research, it is impossible to block until a key appears on Redis and/or Memcached. For example, here is one implementation: https://github.com/sqlalchemy/dogpile.cache/blob/dbf57e4a7b8f4061b735be8b24bbb880fb75802f/dogpile/cache/backends/memcached.py#L42

As you can see, the "locking" works by creating another key and then sleeping on it. No bueno. Something like this https://github.com/mailgun/groupcache needs to be used but before that https://github.com/mailgun/groupcache/issues/9 should be implemented.

**Comment by GiedriusS** [comment]
> From a functional point of view - I'm confident that this change does what it says it does. I do still think we are missing some benchmarks to make it obvious beyond any reasonable doubt that the increase in complexity is worth the performance boost.
> 
> From a maintainability perspective, I think there are still some improvements to be made.
> 
> This is a non-trivial bit of code that has taken me ~30 mins on both reviews to understand and conceptualise how it fits into the wider read & write paths. If the code is easy to understand, people will maintain it and vice versa - what [if Giedrus was hit by a bus](https://legacy.python.org/search/hypermail/python-1994q2/1040.html)?
> 
> Personally, I would like to see extra context in the comments and _ideally_ a sequence diagram that explains why we are making this change.
> 
> I'm not going to block this PR as I'm still pretty inexperienced in the code base, but that is my 2c

Thanks for your review! I will try adding those things. Coming back to it, I still cannot see a way of implementing this in some other way besides working with `context.Context` unless we'd rework a whole lot of interfaces. It's nice ATM that they are generic, it means that we can plug them in wherever we want to. Also, all of this new logic is hidden from the caller's perspective so I would argue that this doesn't introduce a lot of complexity.

**Comment by GiedriusS** [comment]
@ianbillett I have added a benchmark of Series() with 4 different setups:

- Singleflight on, unlimited throughput
- Singleflight off, unlimited throughput
- Singleflight on, 2MiB/s throughput
- Singleflight off, 2MiB/s throughput

The results are pretty good :scream: In some cases, the difference is 10-100x on small throughput benchmarks:

(`/bin/go test -benchmem -run='^$' -benchtime 20x -timeout 9999s -bench '^(BenchmarkBucketSeries)$' github.com/thanos-io/thanos/pkg/store &>bench_data`)

<details>
 <summary>Benchmark results</summary>

```
goos: linux
goarch: amd64
pkg: github.com/thanos-io/thanos/pkg/store
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=true,Throughput=0/1of1000000-16         	      20	  91872278 ns/op	62053467 B/op	    9697 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=true,Throughput=0/10of1000000-16        	      20	  90943501 ns/op	62068092 B/op	    9827 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=true,Throughput=0/1000000of1000000-16   	      20	 856642356 ns/op	1249098976 B/op	10045235 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=true,Throughput=2.097152e+06/1of1000000-16         	      20	4459123241 ns/op	62086337 B/op	    9700 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=true,Throughput=2.097152e+06/10of1000000-16        	      20	4455079743 ns/op	62085030 B/op	    9828 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=true,Throughput=2.097152e+06/1000000of1000000-16   	      20	5193180212 ns/op	1247463237 B/op	10045257 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=false,Throughput=0/1of1000000-16                   	      20	  84673427 ns/op	62106584 B/op	    9768 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=false,Throughput=0/10of1000000-16                  	      20	  87083260 ns/op	62115551 B/op	    9872 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=false,Throughput=0/1000000of1000000-16             	      20	 841538409 ns/op	1274593827 B/op	10056803 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=false,Throughput=2.097152e+06/1of1000000-16        	      20	4456800747 ns/op	62144327 B/op	    9786 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=false,Throughput=2.097152e+06/10of1000000-16       	      20	4468411467 ns/op	62141646 B/op	    9896 allocs/op
BenchmarkBucketSeries/1000000SeriesWith1Samples,Singleflight=false,Throughput=2.097152e+06/1000000of1000000-16  	      20	7576618808 ns/op	1275337376 B/op	10060593 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=true,Throughput=0/1of10000000-16                  	      20	   6576437 ns/op	 4872760 B/op	    1133 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=true,Throughput=0/100of10000000-16                	      20	   6357284 ns/op	 4863187 B/op	    1164 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=true,Throughput=0/10000000of10000000-16           	      20	  82506036 ns/op	128755490 B/op	 1010018 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=true,Throughput=2.097152e+06/1of10000000-16       	      20	 483283161 ns/op	 4899822 B/op	    1146 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=true,Throughput=2.097152e+06/100of10000000-16     	      20	 482696981 ns/op	 4911803 B/op	    1181 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=true,Throughput=2.097152e+06/10000000of10000000-16         	      20	 560929529 ns/op	123472229 B/op	 1009873 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=false,Throughput=0/1of10000000-16                          	      20	   7243218 ns/op	 4898300 B/op	    1167 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=false,Throughput=0/100of10000000-16                        	      20	   6538936 ns/op	 4904600 B/op	    1201 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=false,Throughput=0/10000000of10000000-16                   	      20	  87463841 ns/op	150081864 B/op	 1021428 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=false,Throughput=2.097152e+06/1of10000000-16               	      20	 499757455 ns/op	 4949073 B/op	    1189 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=false,Throughput=2.097152e+06/100of10000000-16             	      20	 500387347 ns/op	 4949694 B/op	    1230 allocs/op
BenchmarkBucketSeries/100000SeriesWith100Samples,Singleflight=false,Throughput=2.097152e+06/10000000of10000000-16        	      20	2971774125 ns/op	152286798 B/op	 1025274 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=true,Throughput=0/1of10000000-16                           	      20	    151952 ns/op	  225324 B/op	     229 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=true,Throughput=0/100of10000000-16                         	      20	    141794 ns/op	  219999 B/op	     227 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=true,Throughput=0/10000000of10000000-16                    	      20	  18142334 ns/op	41465792 B/op	  186320 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=true,Throughput=2.097152e+06/1of10000000-16                	      20	  72945077 ns/op	  238615 B/op	     243 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=true,Throughput=2.097152e+06/100of10000000-16              	      20	  73019216 ns/op	  242035 B/op	     243 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=true,Throughput=2.097152e+06/10000000of10000000-16         	      20	 114852266 ns/op	41475730 B/op	  186373 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=false,Throughput=0/1of10000000-16                          	      20	    164981 ns/op	  258854 B/op	     269 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=false,Throughput=0/100of10000000-16                        	      20	    161352 ns/op	  262310 B/op	     269 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=false,Throughput=0/10000000of10000000-16                   	      20	  29896776 ns/op	151055664 B/op	  229318 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=false,Throughput=2.097152e+06/1of10000000-16               	      20	  88654022 ns/op	  272788 B/op	     292 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=false,Throughput=2.097152e+06/100of10000000-16             	      20	  88854959 ns/op	  284258 B/op	     291 allocs/op
BenchmarkBucketSeries/1SeriesWith10000000Samples,Singleflight=false,Throughput=2.097152e+06/10000000of10000000-16        	      20	8996025188 ns/op	156873714 B/op	  243499 allocs/op
PASS
ok  	github.com/thanos-io/thanos/pkg/store	1202.814s
```
</details>

Being network-bound is the main limiting factor ATM in Thanos performance IMHO so this seems to me like a very important thing to have. Please see some of the rationales here: https://github.com/thanos-io/thanos/pull/4379#discussion_r681655753. Given this, I'd like to merge this functionality for the in-memory cache and then we could move to the more advanced case where there are multiple Store instances.

File: pkg/cache/inmemory.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (kakkoyun) [reply pkg/cache/inmemory.go:314]: Off-topic: My attempt to add groupcache failed because of the drastic changes that needs to be introduced to cache interfaces. The `Get` issue actually would help to solve this. It would be amazing to migrate groupcache and simplify what we have (we can drop memcached support when we move).

**Comment by GiedriusS** [comment]
BTW, don't merge this yet:
* I will remove `time.Sleep` from tests;
* I have also noticed an edge-case that isn't covered - what if Store() never comes? Then a key will be left dangling forever. Let me also improve the code to cover this case :)

**Comment by kakkoyun** [comment]
@GiedriusS I have converted to draft, just in case to prevent accidental merges.

**Comment by matej-g** [comment]
What's the status of this @GiedriusS? Looks like this was nearly finished but then marked stale? Would be a shame to leave this unmerged it seems!

**Comment by GiedriusS** [comment]
We have groupcache now which does "single-flight" so I don't think this is needed anymore.
